pjrt/convert/trace_container.zig

```diff
@@ -0,0 +1,211 @@
+const std = @import("std");
+const trace_events_proto = @import("//tsl:trace_events_proto");
+const xplane_proto = @import("//tsl:xplane_proto");
+const xplane_schema = @import("utils/xplane_schema.zig");
+const xplane_utils = @import("utils/xplane_utils.zig");
+const xplane_visitor = @import("utils/xplane_visitor.zig");
+
+// Constants used as trace_viewer PID (device_id in trace_events.proto).
+// PID 0 is unused.
+// Support up to 500 accelerator devices.
+const first_device_id = 1;
+const last_device_id = 500;
+// Support Upto 200 custom planes as fake devices (i.e., planes with a
+// "/custom:" prefix). See `<project_name>::custom_plane_prefix` for more
+// information
+const first_custom_plane_device_id = last_device_id + 1;
+const max_custom_plane_devices_per_host = 200;
+const last_custom_plane_device_id = first_custom_plane_device_id + max_custom_plane_devices_per_host - 1;
+
+// Host threads are shown as a single fake device.
+pub const host_threads_device_id = last_custom_plane_device_id + 1;
+
+pub const xla_async_op_line_name = "Async XLA Ops";
+
+pub const host_threads_plane_name = "/host:CPU";
+pub const gpu_plane_prefix = "/device:GPU:";
+pub const tpu_plane_prefix = "/device:TPU:";
+pub const custom_plane_prefix = "/device:CUSTOM:";
+
+pub fn GetOrPutResult(comptime Key: type, comptime Value: type) type {
+    return struct {
+        key_ptr: *Key,
+        value_ptr: *Value,
+    };
+}
+
+pub const TraceContainer = struct {
+    metadata: Trace = .{},
+    events: std.ArrayListUnmanaged(*TraceEvent) = .{},
+
+    pub const Trace = struct {
+        devices: std.AutoArrayHashMapUnmanaged(u32, Device) = .{},
+        trace_events: std.ArrayListUnmanaged(TraceEvent) = .{},
+    };
+
+    pub const Device = struct {
+        name: []const u8 = &[_]u8{},
+        device_id: u32 = 0,
+        resources: std.AutoArrayHashMapUnmanaged(u32, trace_events_proto.Resource) = .{},
+    };
+
+    pub const TraceEvent = struct {
+        device_id: u32 = 0,
+        resource_id: u32 = 0,
+        name: []const u8 = &[_]u8{},
+        timestamp_ps: u128 = 0,
+        duration_ps: u64 = 0,
+        args: std.StringArrayHashMapUnmanaged([]const u8) = .{},
+    };
+
+    fn buildDeviceAndResources(allocator: std.mem.Allocator, device_id: u32, plane: *const xplane_visitor.XPlaneVisitor, device: *Device) !void {
+        device.name = plane.name();
+        device.device_id = device_id;
+        const sort_by_ordinal = (device_id == host_threads_device_id);
+        var ordinal: u32 = 0;
+        for (plane.plane.lines.items) |*l| {
+            var line = xplane_visitor.XLineVisitor.init(plane, l);
```

Comment:
here instead of creating a "XLineVisitor" then calling displayeId and displayName on it, can you create a `fn xLineDisplayId(plane, line)` and `fn xLineDisplayName(plane, line)` and call that ?